### ISSUE TITLE
write_bulk and write_interrupt speed

### ISSUE THREAD

**Issue by flukejones** [a1ien/rusb#21]
It seems no matter what I do here, writing takes between 500u to 1ms (64bytes). Which is far far too slow for what I need. The same data written by whatever means I've captured in Windows with wireshark is sub-nanos.

Context: I'm reverse engineering a usb protocol to control my keyboard leds. I'm using rusb to write packets, and the write is too slow.

I'm unsure if this is a libusb issue, rusb, or other. Device speed is reported and "Full".

Adding some `#[inline]` to some parts of rusb and my own code helps by 100-200u  sort of.

**Comment by a1ien** [comment]
>  The same data written by whatever means I've captured in Windows with wireshark is sub-nanos.

What you use to write? How you measure? Also note how many device connect to same hub. They divide time. And I think it worth check speed with just libusb code.

**Comment by flukejones** [comment]
The write in Windows is done with Asus Armoury Crate (ASUS laptop), I think there are three devices.
Measurement is just a Duration check before/after the write call.

I'm not able to check with C, I don't know it :(

**Comment by a1ien** [comment]
All what rusb crate do it's just wrapper to libusb call with minor check.
write_interrupt direct call libusb_interrupt_transfer
https://github.com/a1ien/rusb/blob/master/src/device_handle.rs#L266-L275

**Comment by flukejones** [comment]
I mean that's where I'm confused. This should be blazing fast, rustc will optimise away a lot of things.

Is libusb1-sys an ffi wrapper, of does it actually link in the libusb lib? I know calls to external libs via ffi tend to be slow, they incur a lot of overhead.

**Comment by a1ien** [comment]
>Is libusb1-sys an ffi wrapper, of does it actually link in the libusb lib?
 It's just ffi interface. Just describe function in libusb.

>I know calls to external libs via ffi tend to be slow, they incur a lot of overhead.
If you write code in C you get same overhead. And Ш think you wrong measure. It's not correct to measure just single transfer in usb bus. Especially on small packages. And when they say slowly it’s nanosecond orders

**Comment by flukejones** [comment]
I must be misunderstanding something in how it works. What would be the best way to transfer say, 500 packets of 64 bytes each?

I'm not wondering if I should be checking a reponse somewhere to see if I can write again, and throwing away packets until then... Although that may end up looking quite erratic...

**Comment by flukejones** [comment]
> All what rusb crate do it's just wrapper to libusb call with minor check.
> write_interrupt direct call libusb_interrupt_transfer
> https://github.com/a1ien/rusb/blob/master/src/device_handle.rs#L266-L275

I did try this, functionally the same.

**Comment by flukejones** [comment]
Bit of an update for you. I tried async swamping the device, and async with a sleep of 500ms (using another usb crate I found) and it looks like the EC has a minimum write time of ~1ms... unless it's an OS restriction here.

When I go back to Windows to check things, it looks like that stuff uses direct access to the device. 

For what I need, the LED matrix in this keyboard is 10x64, which really means it takes a minimum of 10ms to write one colour change, and rusb works perfectly for everything there.

In my testing though, it is definitely worth dropping a `#[inline]` above all public functions as it lets rustc choose to inline when desirable. From memory rustc doesn't do this for external crates unless explicit with that tag. I get 200-300 microseconds taken off the transfer time and that's enough to let me even things out a bit.

Thanks for your time and patience while I worked through this :)

**Comment by flukejones** [comment]
`bInterval 1`

```
If this endpoint is of type interrupt, this value is the interval setting for the endpoint—that is, the time between interrupt requests for the endpoint. The value is represented in milliseconds.
```

:facepalm:

**Comment by flukejones** [comment]
I think I worked out what I was seeing in Windows. It's likely to be async transfers using `libusb_fill_bulk_transfer` and `libusb_submit_transfer`.

I notice `libusb_fill_bulk_transfer` is not in `libusb1-sys`. I'll raise issue there.

**Comment by a1ien** [comment]
https://github.com/a1ien/rusb/commit/db0041b87ed119f7a3372cb414228bdad9d1633f#diff-8326cc5f7f20ae752e139771b46c0923R55-R63
Here how you can implement libusb_fill_bulk_transfer in rust

**Comment by flukejones** [comment]
I took some time and re-implemented that removed async commit. And yeah it is not good (there are many issues with the implementation).

It doesn't perform as I wanted to for timings either, seems to be very much the same as one transfer per frame. It's possible I'm looking for something that libusb can't do, where the vendor has custom hid implementation perhaps.

**Comment by a1ien** [comment]
Maybe you wrong use? You must create one `AsyncGroup` and then create 500 `Transfer` and call 500 `submit` after you can call `wait_any`

**Comment by flukejones** [comment]
I'm using a single group for 10 transfers followed by loop on wait_any. It takes close to same time as for 10 single requests.

What I was hoping for was transfer ten in 2ms or less but it ends up being ten in 16 to 20ms. Some appear to complete in 300 microseconds but then others take 2ms...

I'll stuff this work in a branch for experimenting on. Seems to be a good use case for testing rusb.

**Comment by flukejones** [comment]
The major issue here was my lack of understanding of the HID descriptor. The device allows only one transfer per frame, and one frame is 1ms (full-speed device, USB 2.0 spec).

I analysed the Windows wireshark data and indeed it is sending at most, 1 transfer per 1ms. I'm sorry I came looking for the wrong things here, async transfers certainly would not help when the constraint is time. However an async "block of transfers" did help since I need to write a block of ten for a particular effect, so the async part is writing that block (taking 10ms) and not blocking the main thread.

Just some info if you are curious.
